### DISCUSSION TITLE
Throttle button clicks?

### DISCUSSION THREAD

**Discussion by pythonormen** [zauberzeug/nicegui#628] [Q&A] [answered]
Newbie question again :) but currently I'm disabling a button after it's pressed so the user cannot spam a resource heavy API query. Then I read in the documentation that there is a throttle parameter? Is that something that can be used instead? 
Tried the following but had no effect. 

Thanks

```py
from nicegui import ui

def notification():
    ui.notify('Hello!')

ui.button('Hello').on('click', handler=notification, args=None, throttle=5.0)

ui.run()
```

**Comment by falkoschindler** [comment]
That's a very good question, @pythonormen!
Actually, that _should_ work. But I can confirm that it doesn't. I'll have to look into it.

**Reply by pythonormen** [reply]
Works on mousedown but not on click, so that will do for me. But would be great if it works on click events as well.

**Comment by falkoschindler** [comment]
Oh! I understand. Let me explain:

The server tells the client which events have been registered.
For each event it also specifies a throttle duration to save bandwidth.
If multiple handlers have been registered for the same event, the server communicates the _lowest_ throttle duration to satisfy every event handler (more events won't hurt, right?).
A button _always_ registers a click handler with zero throttling for the `on_click` parameter, even if this parameter is not used.
That's why all other click handlers will observe zero throttling.

I see two possible solutions:

1. Let the button only register for the click event if `on_click` is set.

    This only requires an additional if-statement in the button initializer.
    But there might still be confusion in scenarios with multiple handlers with different throttle durations.

2. Introduce another check on the server side if a specific event handler should be called or throttled.

    This would solve the problem in a more generic way.
    But I fear it might allow for timing edge cases where too many events get dropped on the server.

**Comment by falkoschindler** [comment]
A third option came into my mind:

3. Switch from an event-type-based to a listener-based protocol for the communication between server and client.

    As mentioned above, there is currently only one message emitted per event type per element (e.g. "click"), independent of the number of listeners. This makes the implementation and API including `args` and `throttle` arguments pretty complicated and hard to predict.
    Instead we could simply treat every listener separately with its own args and trottle duration. This might cause slightly more traffic, but every event handler gets exactly what it subscribed for. If bandwidth is a concern, the user has the possibility to reduce the number of event handlers in Python with a one-to-one equivalence to the transmitted messages.

This is currently my prefered solution. It probably simplifies code, generates a more predictable behavior, and gives full control to the user. An increase in bandwidth will only occur in rare cases, but then it will be reasonable and easy to avoid.

**Reply by pythonormen** [reply]
Really admire your enthusiasm, but maybe not worth the effort. I'm happy it works on mousedown 😊

**Reply by falkoschindler** [reply]
Sure, in your specific case using mousedown is just fine. But we're always trying to improve the architecture and usability - until it feels _nice_. 😉

**Comment by falkoschindler** [comment, marked as answer]
In [a07641a](https://github.com/zauberzeug/nicegui/commit/a07641a29d932c8e20e9b97ef5fe3e38a3ab8765) I made sure to only subscribe to the click event if an `on_click` handler is given. This should fix the original problem of throttling button clicks.

For the more general rework of the event protocol I created a new issue: #644.

**Reply by pythonormen** [reply]
What does this mean? That the .on 'click' -variant will consider the throttle value after this change? 😊

**Reply by falkoschindler** [reply]
Yes. After this change the button in your example will only register _one_ event handler (not the default `on_click`). So there is no handler with zero throttling, thus the throttle value of `5.0` will be effective.
